tlatools/org.lamport.tlatools/test/tlc2/debug/GetScopedIdentifiersTests.java

```diff
@@ -0,0 +1,95 @@
+
+import tla2sany.semantic.ModuleNode;
+import tla2sany.st.Location;
+
+/**
+ * {@link TLCBreakpointExpression#getScopedIdentifiers(ModuleNode, Location)}
+ * tests, to check that various nested TLA+ syntax constructs have their
+ * introduced identifiers appropriately recorded.
+ */
+public class GetScopedIdentifiersTests {
+
+  private static class TestCase {
+    public final String input;
+    public final String terminator;
+    public final Set<String> expected;
+    public TestCase(String input, String terminator, String... expected) {
+      this.input = input;
+      this.terminator = terminator;
+      this.expected = new HashSet<String>(List.of(expected));
+    }
+  }
+
+  /**
+   * A wrapper in which to interpolate test cases to form a valid module.
+   */
+  private static final String wrapper =
+    "---- MODULE Test ----\n"
+    + "VARIABLE x\n"
+    + "CONSTANT y\n"
+    + "%s\n"
+    + "TRUE%s\n"
+    + "====";
+
+  /**
+   * The location of the expression TRUE in the wrapper. This assumes that
+   * each test case input is a single line.
+   */
+  private static final Location location
+    = new Location(UniqueString.of("Test"), 5, 1, 5, 2);
+
+  /**
+   * A set of test cases for finding scoped identifiers. Each input should
+   * consist of an incomplete operator definition & expression that can be
+   * terminated by expression TRUE located at the start of the next line,
+   * along with a variable-length list of identifiers accessible at the
+   * point of the TRUE expression. A terminator to be appended after TRUE is
+   * also accepted - for example ⟩ or ) or ] or }.
+   */
+  private static final TestCase[] testCases = new TestCase[] {
+    new TestCase("op ≜", ""),
+    new TestCase("op(i, j, k) ≜", "", "j", "i", "k"),
+    new TestCase("op(i) ≜ ∀ j ∈ {} :", "", "i", "j"),
+    new TestCase("op ≜ ∀ i, j ∈ {} :", "", "i", "j"),
+    new TestCase("op ≜ ∀ ⟨i, j, k⟩ ∈ {} :", "", "i", "j", "k"),
+    new TestCase("op(i, j) ≜ LET k == TRUE l == TRUE IN", "", "i", "j", "k", "l"),
+    new TestCase("op ≜ LET i(j, k) == ", " IN TRUE", "i", "j", "k"),
+    new TestCase("op ≜ LET RECURSIVE i i == TRUE IN", "", "i"),
+    new TestCase("op ≜ LET i == TRUE j ==", " IN TRUE", "i", "j"),
+    new TestCase("op(i) ≜ [j, k ∈ {} ↦", "]", "i", "j", "k"),
+    new TestCase("op ≜ [⟨i, j, k⟩ ∈ {} ↦", "]", "i", "j", "k"),
+    new TestCase("op(i, j) ≜ {k ∈ {} :", "}", "i", "j", "k"),
+    new TestCase("RECURSIVE op(_, _) op(f(_,_), i) ≜ op(LAMBDA j, k : ", ", TRUE)", "f", "i", "j", "k"),
+    // This is a bug that needs to be fixed! More likely this will get fixed
+    // when support is added to the semantic checker to introduce "ghost"
+    // expressions at arbitrary points in the semantic tree, rendering all
+    // of this functionality redundant.
+    new TestCase("op(_+_) ≜", "", "+"),
+  };
+
+  /**
+   * Run all the test cases and check the identifiers are found as expected.
+   */
+  @Test
+  public void test() {
```

Comment:
Let's stick to ordinary junit tests and one @Test method per expression so that we know which one fails (if any).

Reply:
This is a deficiency in JUnit 4 (which we use) that was fixed in JUnit 5 with the addition of [dynamic tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dynamic-tests), so each of the test cases here could be generated as a separate test at runtime. Testing has broadly moved away from hardcoded tests toward tests-as-data or even auto-generated test cases. I find it much easier to add another line in the array compared to adding another test method.

Reply:
JUnit 4 supports [parameterized tests](https://github.com/junit-team/junit4/wiki/Parameterized-tests), which is the best of both worlds:
 - You can have a simple array of test inputs
 - Each input is treated as a separate test in JUnit's reports

Reply:
+1 for parameterized tests, which have already been used in e.g. https://github.com/tlaplus/tlaplus/blob/master/tlatools/org.lamport.tlatools/test/tlc2/value/impl/SubsetEnumeratorTest.java